src/main/java/net/javacrumbs/completionstage/CallbackRegistry.java

```diff
@@ -15,22 +15,19 @@
  */
 package net.javacrumbs.completionstage;
 
-import java.util.LinkedList;
 import java.util.Objects;
-import java.util.Queue;
 import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
 
 /**
- * Registry for Consumer callbacks. Works as a state machine switching between NewEmpty, New, Success and Failure state.
+ * Registry for Consumer callbacks. Works as a state machine switching between Initial, Intermediate, Final(Success|Failure) state.
  * <p/>
  * <p>Inspired by {@code org.springframework.util.concurrent.ListenableFutureCallbackRegistry} and
  * {@code com.google.common.util.concurrent.ExecutionList}</p>
  */
 final class CallbackRegistry<T> {
```

Comment:
I have a really, really stupid [performance test](https://github.com/lukas-krecan/completion-stage/blob/master/src/test/java/net/javacrumbs/completionstage/load/LoadTest.java) that just measures behaviour in non-contended scenario and it seems that the old synchronized solution had much better performance (the difference is quite huge so I feel comfortable even though the test is really bad). 

I may be wrong, but since Completable futures are able to hold only one value, I assume that there will not be much lock contention in real-world scenarios neither. Unless you have some better benchmarks, I would prefer to keep the old synchronized solution.